src/main/java/hudson/tasks/SMTPAuthentication.java

```diff
@@ -1,40 +1,131 @@
 package hudson.tasks;
 
+import com.cloudbees.plugins.credentials.*;
+import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.domains.Domain;
+import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
 import hudson.Extension;
 import hudson.model.AbstractDescribableImpl;
 import hudson.model.Descriptor;
+import hudson.model.Item;
+import hudson.security.ACL;
+import hudson.util.ListBoxModel;
 import hudson.util.Secret;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.AncestorInPath;
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.util.Collections;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.StreamSupport;
 
 /**
  * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
  */
 public class SMTPAuthentication extends AbstractDescribableImpl<SMTPAuthentication> {
 
-    private String username;
+    private static final Logger LOGGER = Logger.getLogger(SMTPAuthentication.class.getName());
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient String username;
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient Secret password;
 
-    private Secret password;
+    /** The ID of the Jenkins Username/Password credential to use. */
+    private String credentialsId;
 
     @DataBoundConstructor
-    public SMTPAuthentication(String username, Secret password) {
+    public SMTPAuthentication(String credentialsId) {
+        this.credentialsId = credentialsId;
+    }
+
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    SMTPAuthentication(String username, Secret password) {
         this.username = username;
         this.password = password;
+        readResolve();
```

Comment:
I've put this here since this legacy constructor should only be called by the migration code in Mailer, and I wanted to avoid constructing credentials in 2 places (`Mailer#readResolve` and `SMTPAuthentication#readResolve`). Is this alright?